src/operator/join.rs

```diff
@@ -4,38 +4,57 @@ use crate::{
         operator_traits::{BinaryOperator, Operator},
         Circuit, Scope, Stream,
     },
+    operator::BinaryOperatorAdapter,
+    SharedRef,
 };
 use std::{borrow::Cow, marker::PhantomData};
 
-impl<P, I1> Stream<Circuit<P>, I1>
+impl<P, SR1> Stream<Circuit<P>, SR1>
 where
     P: Clone + 'static,
 {
     /// Apply [`Join`] operator to `self` and `other`.
     ///
     /// See [`Join`] operator for more info.
-    pub fn join<K, V1, V2, V, W, F, I2, Z>(
+    pub fn join<K, V1, V2, V, W, F, SR2, Z>(
         &self,
-        other: &Stream<Circuit<P>, I2>,
+        other: &Stream<Circuit<P>, SR2>,
         f: F,
     ) -> Stream<Circuit<P>, Z>
     where
         K: KeyProperties,
         V1: KeyProperties,
         V2: KeyProperties,
         W: ZRingValue,
-        I1: IndexedZSet<K, V1, W>,
-        for<'a> &'a I1: IntoIterator<Item = (&'a K, &'a I1::ZSet)>,
-        for<'a> &'a I1::ZSet: IntoIterator<Item = (&'a V1, &'a W)>,
-        I2: IndexedZSet<K, V2, W>,
-        for<'a> &'a I2: IntoIterator<Item = (&'a K, &'a I2::ZSet)>,
-        for<'a> &'a I2::ZSet: IntoIterator<Item = (&'a V2, &'a W)>,
+        SR1: SharedRef + 'static,
+        <SR1 as SharedRef>::Target: IndexedZSet<K, V1, W>,
+        for<'a> &'a <SR1 as SharedRef>::Target: IntoIterator<
+            Item = (
+                &'a K,
+                &'a <<SR1 as SharedRef>::Target as IndexedZSet<K, V1, W>>::ZSet,
+            ),
+        >,
+        for<'a> &'a <<SR1 as SharedRef>::Target as IndexedZSet<K, V1, W>>::ZSet:
+            IntoIterator<Item = (&'a V1, &'a W)>,
+        SR2: SharedRef + 'static,
+        <SR2 as SharedRef>::Target: IndexedZSet<K, V2, W>,
+        for<'a> &'a <SR2 as SharedRef>::Target: IntoIterator<
+            Item = (
+                &'a K,
+                &'a <<SR2 as SharedRef>::Target as IndexedZSet<K, V2, W>>::ZSet,
+            ),
+        >,
+        for<'a> &'a <<SR2 as SharedRef>::Target as IndexedZSet<K, V2, W>>::ZSet:
+            IntoIterator<Item = (&'a V2, &'a W)>,
```

Comment:
Yikes

Reply:
I know, it keeps getting better.

Reply:
I wonder if there is a Guinness record for the longest chain of constraints we can apply for.